src/hashstore/filehashstore.py

```diff
@@ -404,77 +412,240 @@ def lookup_algo(algo):
-        (
-            additional_algorithm_checked,
-            checksum_algorithm_checked,
-        ) = self._validate_algorithms_and_checksum(
-            additional_algorithm, checksum, checksum_algorithm
+            "FileHashStore - tag_object: Tagging object cid: {%s} with pid: {%s}.",
+            cid,
+            pid,
         )
-
-        # Wait for the pid to release if it's in use
-        while pid in self.object_locked_pids:
+        self._check_string(pid, "pid", "tag_object")
+        self._check_string(cid, "cid", "tag_object")
+        # Wait for the cid to release if it's being tagged
+        while cid in self.reference_locked_cids:
             logging.debug(
-                "FileHashStore - store_object: %s is currently being stored. Waiting.",
-                pid,
+                "FileHashStore - tag_object: (cid) %s is currently locked. Waiting.",
+                cid,
             )
             time.sleep(self.time_out_sec)
-        # Modify object_locked_pids consecutively
-        with self.object_lock:
+        # Modify reference_locked_cids consecutively
+        with self.reference_lock:
             logging.debug(
-                "FileHashStore - store_object: Adding pid: %s to object_locked_pids.",
-                pid,
+                "FileHashStore - tag_object: Adding cid: %s to reference_locked_cids.",
+                cid,
             )
-            self.object_locked_pids.append(pid)
+            self.reference_locked_cids.append(cid)
         try:
-            logging.debug(
-                "FileHashStore - store_object: Attempting to store object for pid: %s",
-                pid,
-            )
-            object_metadata = self.put_object(
-                pid,
-                data,
-                additional_algorithm=additional_algorithm_checked,
-                checksum=checksum,
-                checksum_algorithm=checksum_algorithm_checked,
-                file_size_to_validate=expected_object_size,
-            )
+            pid_ref_abs_path = self.resolve_path("pid", pid)
+            cid_ref_abs_path = self.resolve_path("cid", cid)
+            tmp_root_path = self.get_store_path("refs") / "tmp"
+
+            # Proceed to tagging process
+            if os.path.exists(pid_ref_abs_path):
+                # A pid reference file can only contain one cid
+                exception_string = (
+                    "FileHashStore - write_pid_refs_file: pid ref file already exists for"
+                    + pid_ref_abs_path
+                )
+                logging.error(exception_string)
+                raise FileExistsError(exception_string)
+            elif os.path.exists(cid_ref_abs_path):
+                # Create the pid refs file
+                pid_tmp_file_path = self._write_pid_refs_file(tmp_root_path, cid)
+                self.create_path(os.path.dirname(pid_ref_abs_path))
+                shutil.move(pid_tmp_file_path, pid_ref_abs_path)
```

Comment:
curious why you create a temp file then move it? The pid files are tiny, right?

Reply:
I followed this pattern so that this process of writing the refs file could be atomic (either the file was written successfully or it's not written at all, and they are tiny indeed!).